scopes/preview/preview/pre-bundle.ts

```diff
@@ -0,0 +1,134 @@
+  const mergedConfig = { ...config, ...windowConfig };
+  __ID_SETTERS__
+  function render(...props) {
+    return Harmony.load(
+      [
+        ...customAspects,
+        __IDENTIFIERS__,
+      ],
+      __RUNTIME_NAME__,
+      mergedConfig
+    ).then((harmony) => {
+      return harmony
+        .run()
+        .then(() => harmony.get(__ROOT_ASPECT__))
+        .then((rootExtension) => {
+          const ssrSetup = !isBrowser && rootExtension.setupSsr;
+          const setup = rootExtension.setup;
+          const setupFunc = (ssrSetup || setup || function noop() {}).bind(
+            rootExtension
+          );
+          return Promise.resolve(setupFunc()).then(
+            () => {
+              return rootExtension
+            }
+          );
+        })
+        .then((rootExtension) => {
+          if (isBrowser) {
+            return rootExtension.render(
+              __ROOT_EXTENSION_NAME__,
+              ...props
+            );
+          } else {
+            return rootExtension.renderSsr(
+              __ROOT_EXTENSION_NAME__,
+              ...props
+            );
+          }
+        })
+        .catch((err) => {
+          throw err;
+        });
+    });
+  }
+  if (isBrowser || __RUNTIME_NAME__ === "main") render();
+};
+`;
+
+export const generatePreBundlePreviewEntry = (
+  aspectDefs: AspectDefinition[],
+  rootExtensionName: string,
+  runtimeName: string,
+  rootAspectId: string,
+  dir: string
+) => {
+  const harmonyImport = createHarmonyImports();
+  const imports = createImports(aspectDefs);
+  const identifiers = getIdentifiers(aspectDefs, 'Aspect');
+  const idSetters = getIdSetters(aspectDefs, 'Aspect');
+  const contents = ENTRY_CONTENT_TEMPLATE.replace('__IMPORTS__', [harmonyImport, imports].join('\n'))
+    .replace('__IDENTIFIERS__', identifiers.join(', '))
+    .replace('__ID_SETTERS__', idSetters.join('\n'))
+    .replaceAll('__RUNTIME_NAME__', JSON.stringify(runtimeName))
+    .replaceAll('__ROOT_ASPECT__', JSON.stringify(rootAspectId))
+    .replaceAll('__ROOT_EXTENSION_NAME__', JSON.stringify(rootExtensionName));
+  const entryPath = resolve(join(dir, `ui-bundle-entry.${sha1(contents)}.js`));
+  if (!fs.existsSync(entryPath)) {
+    fs.outputFileSync(entryPath, contents);
+  }
+  return entryPath;
+};
+
+export async function buildPreBundlePreview(resolvedAspects: AspectDefinition[], customOutputDir?: string) {
+  const outputDir = customOutputDir || resolve(PUBLIC_DIR);
+  const mainEntry = generatePreBundlePreviewEntry(
+    resolvedAspects,
+    UIROOT_ASPECT_ID,
+    RUNTIME_NAME,
+    PreviewAspect.id,
+    __dirname
+  );
+  const config = createWebpackConfig(outputDir, mainEntry);
+  const compiler = webpack(config);
+  const compilerRun = promisify(compiler.run.bind(compiler));
+  const results = await compilerRun();
+  if (!results) throw new Error();
+  if (results?.hasErrors()) {
+    clearConsole();
+    throw new Error(results?.toString());
```

Comment:
not sure printing the entire results here is a good idea. it can be hugs.. 
We have some place in the webpack aspect that we kind of concat all the errors into one error. you can use the same. let me know if you can't find it

Reply:
I see. It's just copied from `ui.main.runtime.ts`

https://github.com/teambit/bit/blob/614778af4d86b83a1eb31a4acaf3bb12d037fc8b/scopes/ui-foundation/ui/ui.main.runtime.ts#L242

Do you think we should change them together?